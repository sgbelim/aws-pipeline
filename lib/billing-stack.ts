import * as cdk from '@aws-cdk/core';
import {Budget} from "./constructs/budget";
import {StackProps} from "@aws-cdk/core";

interface BillilngStackProps extends StackProps {
    budgetAmount: number,
    emailAddress: string
}

export class BillingStack extends cdk.Stack {
    constructor(scope: cdk.Construct, id: string, props: BillilngStackProps) {
        super(scope, id, props);

        new Budget(this, 'Budget', {
            budgetAmount: props.budgetAmount,
            emailAddress: props.emailAddress
        })
    }
}
