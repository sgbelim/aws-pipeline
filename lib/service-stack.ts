import * as cdk from '@aws-cdk/core';
import {CfnParametersCode, Code, Function, Runtime} from "@aws-cdk/aws-lambda";
import {HttpApi} from "@aws-cdk/aws-apigatewayv2";
import {LambdaProxyIntegration} from "@aws-cdk/aws-apigatewayv2-integrations";
import {CfnParameter} from "@aws-cdk/core";

export class ServiceStack extends cdk.Stack {

    public readonly serviceCode: CfnParametersCode;

    constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);

        this.serviceCode = Code.fromCfnParameters({
            bucketNameParam: new CfnParameter(this, 'ServiceLambdaBucket'),
            objectKeyParam: new CfnParameter(this, 'ServiceLambdaBucketObject'),
        });

        const lambda = new Function(this, 'ServiceLambda', {
            runtime: Runtime.NODEJS_14_X,
            handler: 'src/lambda.handler',
            code: this.serviceCode,
            functionName: 'ServiceLambda'
        })

        new HttpApi(this, 'ApiGatewayService', {
            defaultIntegration: new LambdaProxyIntegration({
                handler: lambda
            }),
            apiName: 'MyService'
        })

    }
}
