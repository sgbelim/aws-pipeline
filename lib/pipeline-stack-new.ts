import * as cdk from '@aws-cdk/core';
import {SecretValue} from '@aws-cdk/core';
import {Artifact, IStage, Pipeline} from "@aws-cdk/aws-codepipeline";
import {
    CloudFormationCreateUpdateStackAction,
    CodeBuildAction,
    CodeBuildActionType,
    GitHubSourceAction
} from "@aws-cdk/aws-codepipeline-actions";
import {BuildEnvironmentVariableType, BuildSpec, LinuxBuildImage, PipelineProject} from "@aws-cdk/aws-codebuild";
import {ServiceStack} from "./service-stack";
import {BillingStack} from "./billing-stack";

export class PipelineStackNew extends cdk.Stack {

    public readonly pipeline: Pipeline;
    public readonly cdkBuildOutput: Artifact;
    public readonly serviceBuildOutput: Artifact;
    public readonly serviceSourceOutput: Artifact;

    constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);

        this.pipeline = new Pipeline(this, 'Pipeline', {
            pipelineName: 'Pipeline',
            crossAccountKeys: false,
            restartExecutionOnUpdate: true
        })

        const cdkSourceOutput = new Artifact('CdkSourceOutput')
        this.serviceSourceOutput = new Artifact('ServiceSourceOutput')

        this.pipeline.addStage({
            stageName: 'Source',
            actions: [
                new GitHubSourceAction({
                    owner: 'sgbelim',
                    repo: 'aws-pipeline',
                    branch: 'main',
                    actionName: 'Pipeline_Source',
                    oauthToken: SecretValue.secretsManager('github-token'),
                    output: cdkSourceOutput
                }),
                new GitHubSourceAction({
                    owner: 'sgbelim',
                    repo: 'express-lambda',
                    branch: 'main',
                    actionName: 'Service_Source',
                    oauthToken: SecretValue.secretsManager('github-token'),
                    output: this.serviceSourceOutput
                })
            ]
        })

        this.cdkBuildOutput = new Artifact('CdkBuildOutput')
        this.serviceBuildOutput = new Artifact('ServiceBuildOutput')

        this.pipeline.addStage({
            stageName: 'Build',
            actions: [
                new CodeBuildAction({
                    actionName: 'CDK_Build',
                    input: cdkSourceOutput,
                    outputs: [this.cdkBuildOutput],
                    project: new PipelineProject(this, 'CdkBuildProject', {
                        environment: {
                            buildImage: LinuxBuildImage.STANDARD_5_0
                        },
                        buildSpec: BuildSpec.fromSourceFilename('build-specs/cdk-build-spec.yml')
                    })
                }),
                new CodeBuildAction({
                    actionName: 'Service_Build',
                    input: this.serviceSourceOutput,
                    outputs: [this.serviceBuildOutput],
                    project: new PipelineProject(this, 'ServiceBuildProject', {
                        environment: {
                            buildImage: LinuxBuildImage.STANDARD_5_0
                        },
                        buildSpec: BuildSpec.fromSourceFilename('build-specs/service-build-spec.yml')
                    })
                })
            ]
        })

        this.pipeline.addStage({
            stageName: 'Pipeline_Update',
            actions: [
                new CloudFormationCreateUpdateStackAction({
                    actionName: "Pipeline_Update",
                    stackName: "PipelineStackNew",
                    templatePath: this.cdkBuildOutput.atPath('PipelineStackNew.template.json'),
                    adminPermissions: true
                })
            ]
        })
    }

    public addServiceStage(serviceStack: ServiceStack, stageName: string): IStage {

        return this.pipeline.addStage({
            stageName: stageName,
            actions: [
                new CloudFormationCreateUpdateStackAction({
                    actionName: "Service_Update",
                    stackName: serviceStack.stackName,
                    templatePath: this.cdkBuildOutput.atPath(`${serviceStack.stackName}.template.json`),
                    adminPermissions: true,
                    parameterOverrides: {
                        ...serviceStack.serviceCode.assign(
                            this.serviceBuildOutput.s3Location
                        )
                    },
                    extraInputs: [this.serviceBuildOutput],
                })
            ]
        })
    }

    public addBillingStackToStage(billingStack: BillingStack, stage: IStage) {

        stage.addAction(
            new CloudFormationCreateUpdateStackAction({
                actionName: "Billing_Update",
                stackName: billingStack.stackName,
                templatePath: this.cdkBuildOutput.atPath(`${billingStack.stackName}.template.json`),
                adminPermissions: true
            })
        )
    }

    public addServiceIntegrationTestToStage(stage: IStage, serviceEndpoint: string) {

        stage.addAction(
            new CodeBuildAction({
                actionName: 'Integration_Tests',
                input: this.serviceSourceOutput,
                project: new PipelineProject(this, 'ServiceIntegrationTestsProject', {
                    environment: {
                        buildImage: LinuxBuildImage.STANDARD_5_0
                    },
                    buildSpec: BuildSpec.fromSourceFilename('build-specs/integration-test-build-spec.yml')
                }),
                environmentVariables: {
                    SERVICE_ENDPOINT: {
                        value: serviceEndpoint,
                        type: BuildEnvironmentVariableType.PLAINTEXT
                    }
                },
                type: CodeBuildActionType.TEST,
                runOrder: 2
            })
        )
    }

}
