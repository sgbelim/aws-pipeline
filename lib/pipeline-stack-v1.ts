import * as cdk from '@aws-cdk/core';
import {Artifact, Pipeline} from "@aws-cdk/aws-codepipeline";
import {
    CloudFormationCreateUpdateStackAction,
    CodeBuildAction,
    GitHubSourceAction
} from "@aws-cdk/aws-codepipeline-actions";
import {SecretValue} from "@aws-cdk/core";
import {BuildSpec, LinuxBuildImage, PipelineProject} from "@aws-cdk/aws-codebuild";
import {ServiceStack} from "./service-stack";

export class PipelineStackV1 extends cdk.Stack {

    public readonly pipeline: Pipeline;
    public readonly cdkBuildOutput: Artifact;
    public readonly serviceBuildOutput: Artifact;

    constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);

        this.pipeline = new Pipeline(this, 'Pipeline', {
            pipelineName: 'Pipeline',
            crossAccountKeys: false,
            restartExecutionOnUpdate: true
        })

        const cdkSourceOutput = new Artifact('CdkSourceOutput')
        const serviceSourceOutput = new Artifact('ServiceSourceOutput')

        this.pipeline.addStage({
            stageName: 'Source',
            actions: [
                new GitHubSourceAction({
                    owner: 'sgbelim',
                    repo: 'aws-pipeline',
                    branch: 'main',
                    actionName: 'Pipeline_Source',
                    oauthToken: SecretValue.secretsManager('github-token'),
                    output: cdkSourceOutput
                }),
                new GitHubSourceAction({
                    owner: 'sgbelim',
                    repo: 'express-lambda',
                    branch: 'main',
                    actionName: 'Service_Source',
                    oauthToken: SecretValue.secretsManager('github-token'),
                    output: serviceSourceOutput
                })
            ]
        })

        this.cdkBuildOutput = new Artifact('CdkBuildOutput')
        this.serviceBuildOutput = new Artifact('ServiceBuildOutput')

        this.pipeline.addStage({
            stageName: 'Build',
            actions: [
                new CodeBuildAction({
                    actionName: 'CDK_Build',
                    input: cdkSourceOutput,
                    outputs: [this.cdkBuildOutput],
                    project: new PipelineProject(this, 'CdkBuildProject', {
                        environment: {
                            buildImage: LinuxBuildImage.STANDARD_5_0
                        },
                        buildSpec: BuildSpec.fromSourceFilename('build-specs/cdk-build-spec.yml')
                    })
                }),
                new CodeBuildAction({
                    actionName: 'Service_Build',
                    input: serviceSourceOutput,
                    outputs: [this.serviceBuildOutput],
                    project: new PipelineProject(this, 'ServiceBuildProject', {
                        environment: {
                            buildImage: LinuxBuildImage.STANDARD_5_0
                        },
                        buildSpec: BuildSpec.fromSourceFilename('build-specs/service-build-spec.yml')
                    })
                })
            ]
        })

        this.pipeline.addStage({
            stageName: 'Pipeline_Update',
            actions: [
                new CloudFormationCreateUpdateStackAction({
                    actionName: "Pipeline_Update",
                    stackName: "PipelineStack",
                    templatePath: this.cdkBuildOutput.atPath('PipelineStack.template.json'),
                    adminPermissions: true
                })
            ]
        })
    }

    public addServiceState(serviceStack: ServiceStack, stageName: string) {

        this.pipeline.addStage({
            stageName: stageName,
            actions: [
                new CloudFormationCreateUpdateStackAction({
                    actionName: "Service_Update",
                    stackName: serviceStack.stackName,
                    templatePath: this.cdkBuildOutput.atPath(`${serviceStack.stackName}.template.json`),
                    adminPermissions: true,
                    parameterOverrides: {
                        ...serviceStack.serviceCode.assign(
                            this.serviceBuildOutput.s3Location
                        ),
                    },
                    extraInputs: [this.serviceBuildOutput],
                })
            ]
        })

    }
}
