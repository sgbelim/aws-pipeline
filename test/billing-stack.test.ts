import {SynthUtils} from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import * as Billing from '../lib/billing-stack';

test('Billing Stack', () => {

    // Arrange
    const app = new cdk.App();

    // Act
    const stack = new Billing.BillingStack(app, 'MyBillingStack', {
        budgetAmount: 1,
        emailAddress: "test@example.com"
    });

    // Assert
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
