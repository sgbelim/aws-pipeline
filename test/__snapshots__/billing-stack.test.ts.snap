// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Billing Stack 1`] = `
Object {
  "Resources": Object {
    "BudgetBudgetNew50BD65B3": Object {
      "Properties": Object {
        "Budget": Object {
          "BudgetLimit": Object {
            "Amount": 1,
            "Unit": "USD",
          },
          "BudgetName": "Monthly Budget",
          "BudgetType": "COST",
          "TimeUnit": "MONTHLY",
        },
        "NotificationsWithSubscribers": Array [
          Object {
            "Notification": Object {
              "ComparisonOperator": "GREATER_THAN",
              "NotificationType": "ACTUAL",
              "Threshold": 100,
              "ThresholdType": "PERCENTAGE",
            },
            "Subscribers": Array [
              Object {
                "Address": "test@example.com",
                "SubscriptionType": "EMAIL",
              },
            ],
          },
        ],
      },
      "Type": "AWS::Budgets::Budget",
    },
  },
}
`;
